src/sbf.cpp

```diff
@@ -0,0 +1,579 @@
+				snprintf(msg, sizeof(msg), SBF_CONFIG_RTCM_STATIC_OFFSET, 0.0, 0.0, 0.0);
+				sendMessageAndWaitForAck(msg, SBF_CONFIG_TIMEOUT);
+				sendMessageAndWaitForAck(SBF_CONFIG_RTCM_STATIC1, SBF_CONFIG_TIMEOUT);
+				sendMessageAndWaitForAck(SBF_CONFIG_RTCM_STATIC2, SBF_CONFIG_TIMEOUT);
+			}
+		}
+
+		break;
+	}
+
+	if (baud_i >= sizeof(baudrates) / sizeof(baudrates[0])) {
+		return -1; // connection and/or baudrate detection failed
+	}
+
+	_configured = true;
+	return 0;
+}
+
+bool
+GPSDriverSBF::sendMessage(const char *msg)
+{
+	SBF_DEBUG("Send MSG: %s", msg);
+
+	// Send message
+	int length = static_cast<int>(strlen(msg));
+
+	if (write(msg, length) != length) {
+		return false;
+	}
+
+	return true;
+}
+
+bool
+GPSDriverSBF::sendMessageAndWaitForAck(const char *msg, const int timeout)
+{
+	SBF_DEBUG("Send MSG: %s", msg);
+
+	// Send message
+	int length = static_cast<int>(strlen(msg));
+
+	if (write(msg, length) != length) {
+		return false;
+	}
+
+	// Wait for acknowledge
+	// For all valid set -, get - and exe -commands, the first line of the reply is an exact copy
+	// of the command as entered by the user, preceded with "$R:"
+	char buf[GPS_READ_BUFFER_SIZE];
+	size_t offset = 0;
+	memset(buf, 0, sizeof(buf));
+	gps_abstime time_started = gps_absolute_time();
+
+	bool found_response = false;
+
+	do {
+		int ret = read(reinterpret_cast<uint8_t *>(buf) + offset, sizeof(buf) - offset, timeout);
+
+		if (ret < 0) {
+			// something went wrong when polling or reading
+			SBF_WARN("sbf poll_or_read err");
+			return false;
+		}
+
+		offset += ret;
+
+		if (!found_response && strstr(buf, "$R: ") != nullptr) {
+			SBF_DEBUG("READ %d: %s", offset, buf);
+			found_response = true;
+		}
+
+		if (offset >= sizeof(buf)) {
+			offset = 0;
+			memset(buf, 0, sizeof(buf));
+		}
+
+	} while (time_started + 1000 * timeout > gps_absolute_time());
+
+	return found_response;
+}
+
+int    // -1 = error, 0 = no message handled, 1 = message handled, 2 = sat info message handled
+GPSDriverSBF::receive(unsigned timeout)
+{
+	// Do not receive messages until we're configured
+	if (!_configured) {
+		usleep(timeout * 1000);
+		return 0;
+	}
+
+	uint8_t buf[GPS_READ_BUFFER_SIZE];
+
+	// timeout additional to poll
+	gps_abstime time_started = gps_absolute_time();
+
+	int handled = 0;
+
+	while (true) {
+		// Wait for only SBF_PACKET_TIMEOUT if something already received.
+		int ret = read(buf, sizeof(buf), handled ? SBF_PACKET_TIMEOUT : timeout);
+
+		if (ret < 0) {
+			// something went wrong when polling or reading
+			SBF_WARN("ubx poll_or_read err");
+			return -1;
+
+		} else {
+			// SBF_DEBUG("Read %d bytes", ret);
+
+			// pass received bytes to the packet decoder
+			for (int i = 0; i < ret; i++) {
+				handled |= parseChar(buf[i]);
+				// SBF_DEBUG("parsed %d: 0x%x", i, buf[i]);
+			}
+		}
+
+		if (handled > 0) {
+			return handled;
+		}
+
+		// abort after 2s if no useful packets received
+		if (time_started + 2000 * 1000 < gps_absolute_time()) {
```

Comment:
Can you use the `timeout` argument here?

Reply:
Yes - original ubx code uses timeout argument. I changed it.